feature/experimental/p4rt/otg_tests/google_discovery_protocol_packetin_test/google_discovery_protocol_packetin_test.go

```diff
@@ -0,0 +1,529 @@
+	p4InfoFile            = flag.String("p4info_file_location", "../../wbb.p4info.pb.txt", "Path to the p4info file.")
+	gdpSrcMAC             = flag.String("gdp_src_MAC", "00:01:00:02:00:03", "source MAC address for PacketIn")
+	streamName            = "p4rt"
+	gdpMAC                = "00:0a:da:f0:f0:f0"
+	gdpEtherType          = *ygot.Uint32(0x6007)
+	deviceID              = *ygot.Uint64(1)
+	portID                = *ygot.Uint32(10)
+	electionID            = *ygot.Uint64(100)
+	METADATA_INGRESS_PORT = *ygot.Uint32(1)
+	METADATA_EGRESS_PORT  = *ygot.Uint32(2)
+)
+
+var (
+	dutPort1 = attrs.Attributes{
+		Desc:    "dutPort1",
+		IPv4:    "192.0.2.1",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	atePort1 = attrs.Attributes{
+		Name:    "atePort1",
+		MAC:     "02:11:01:00:00:01",
+		IPv4:    "192.0.2.2",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	dutPort2 = attrs.Attributes{
+		Desc:    "dutPort2",
+		IPv4:    "192.0.2.5",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	atePort2 = attrs.Attributes{
+		Name:    "atePort2",
+		MAC:     "02:12:01:00:00:01",
+		IPv4:    "192.0.2.6",
+		IPv4Len: ipv4PrefixLen,
+	}
+)
+
+type PacketIO interface {
+	GetTableEntry(delete bool) []*p4rtutils.ACLWbbIngressTableEntryInfo
+	GetPacketTemplate() *PacketIOPacket
+	GetTrafficFlow(ate *ondatra.ATEDevice, frameSize uint32, frameRate uint64) []gosnappi.Flow
+	GetEgressPort() []string
+	GetIngressPort() string
+}
+
+type PacketIOPacket struct {
+	SrcMAC, DstMAC *string
+	EthernetType   *uint32
+}
+
+type testArgs struct {
+	ctx      context.Context
+	leader   *p4rt_client.P4RTClient
+	follower *p4rt_client.P4RTClient
+	dut      *ondatra.DUTDevice
+	ate      *ondatra.ATEDevice
+	top      gosnappi.Config
+	packetIO PacketIO
+}
+
+// programmTableEntry programs or deletes p4rt table entry based on delete flag.
+func programmTableEntry(ctx context.Context, t *testing.T, client *p4rt_client.P4RTClient, packetIO PacketIO, delete bool) error {
+	t.Helper()
+	err := client.Write(&p4_v1.WriteRequest{
+		DeviceId:   deviceID,
+		ElectionId: &p4_v1.Uint128{High: uint64(0), Low: electionID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			packetIO.GetTableEntry(delete),
+		),
+		Atomicity: p4_v1.WriteRequest_CONTINUE_ON_ERROR,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// decodePacket decodes L2 header in the packet and returns destination MAC and ethernet type.
+func decodePacket(t *testing.T, packetData []byte) (string, layers.EthernetType) {
+	t.Helper()
+	packet := gopacket.NewPacket(packetData, layers.LayerTypeEthernet, gopacket.Default)
+	etherHeader := packet.Layer(layers.LayerTypeEthernet)
+	if etherHeader != nil {
+		header, decoded := etherHeader.(*layers.Ethernet)
+		if decoded {
+			return header.DstMAC.String(), header.EthernetType
+		}
+	}
+	return "", layers.EthernetType(0)
+}
+
+// testTraffic sends traffic flow for duration seconds.
+func testTraffic(t *testing.T, top gosnappi.Config, ate *ondatra.ATEDevice, flows []gosnappi.Flow, srcEndPoint gosnappi.Port, duration int) {
```

Comment:
testTraffic will need to return the count of packets for the flow after the update.